### ISSUE TITLE
fromJson intermittently fails with IllegalStateException when deserializing recursive structures in parallel

### ISSUE THREAD

**Issue by mwalkerr** [google/gson#1829]
We are seeing an issue where `fromJson` results in a `JsonSyntaxException` with the root cause being an `IllegalStateException`. The stacktrace that comes from within the GSON library is:
```
Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException
at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226)
at com.google.gson.Gson.fromJson(Gson.java:927)
at com.google.gson.Gson.fromJson(Gson.java:892)
at com.google.gson.Gson.fromJson(Gson.java:841)
at com.google.gson.Gson.fromJson(Gson.java:813)
at com.stripe.net.LiveStripeResponseGetter.request(LiveStripeResponseGetter.java:66)
... 136 common frames omitted
Caused by: java.lang.IllegalStateException: null
at com.google.gson.Gson$FutureTypeAdapter.read(Gson.java:1009)
at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
... 141 common frames omitted
```
This seems to be due to a race condition in `Gson.getAdapter` where a `FutureTypeAdapter` is created by one thread and accessed by another before the delegate is set
https://github.com/google/gson/blob/5370b076e329444860b0ea0c7f33b10504281f27/gson/src/main/java/com/google/gson/Gson.java#L433

edit: This looks to be well described by the comment on #625 : https://github.com/google/gson/issues/625#issuecomment-321373437

**Comment by Marcono1234** [comment]
This appears to be the same issue as #625

**Comment by mwalkerr** [comment]
> This appears to be the same issue as #625

@Marcono1234 I believe you're right, thank you for linking that. Wish I would have found that sooner!

**Comment by Marcono1234** [comment]
This should be fixed by #1832, which will be available in the next release
